quickcheck-dynamic/src/Test/QuickCheck/DynamicLogic/Quantify.hs

```diff
@@ -32,12 +32,21 @@ import Test.QuickCheck
 import Test.QuickCheck.DynamicLogic.CanGenerate
 import Test.QuickCheck.StateModel
 
--- | A `Quantification` over a type @a@ is a generator that can be used with
---   `Plutus.Contract.Test.ContractModel.forAllQ` to generate random values in
---   DL scenarios. In addition to a QuickCheck generator a `Quantification` contains a shrinking
---   strategy that ensures that shrunk values stay in the range of the generator.
+-- | A `Quantification` over a type @a@ is a generator that can be used to generate random values in
+-- DL scenarios.
+--
+-- A `Quantification` is similar to a  `Test.QuickCheck.Arbitrary`, it groups together:
+--
+-- * A standard QuickCheck _generator_ in the `Gen` monad,
+-- * A _shrinking_ strategy for generated values in the case of a
+--   failures ensuring they stay within the domain,
+-- * A _predicate_ allowing finer grained control on generation
+--   and shrinking process, e.g in the case the range of the generator
+--   depends on trace context.
 data Quantification a = Quantification
-  { genQ :: Maybe (Gen a)
+  { -- FIXME: why is this a Maybe? It makes things annoying down the road, smells like
+    -- something which would benefit from better typing?
+    genQ :: Maybe (Gen a)
```

Comment:
This is to allow empty generators. A problem that you often run into with QuickCheck is `elements` crashing on an empty list. Having a `Maybe` here allows `elements` to succeed (but with an empty generator).

This lets you write nice code like this (from the registry example):
```
pickAlive :: DL RegState (Var ThreadId)
pickAlive = do
  alive <- getAlive
  forAllQ $ elementsQ alive
```
There is no need to check if there are any alive threads here. If there aren't, we'll get an empty generator and the DL machinery will prune that branch.